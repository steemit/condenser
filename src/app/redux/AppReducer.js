import { Map, OrderedMap } from 'immutable';
import tt from 'counterpart';

const STEEM_API_ERROR = 'app/STEEM_API_ERROR';
const FETCH_DATA_BEGIN = 'app/FETCH_DATA_BEGIN';
const FETCH_DATA_END = 'app/FETCH_DATA_END';
const ADD_NOTIFICATION = 'app/ADD_NOTIFICATION';
const REMOVE_NOTIFICATION = 'app/REMOVE_NOTIFICATION';
const UPDATE_NOTIFICOUNTERS = 'app/UPDATE_NOTIFICOUNTERS';
export const SET_USER_PREFERENCES = 'app/SET_USER_PREFERENCES';
export const TOGGLE_NIGHTMODE = 'app/TOGGLE_NIGHTMODE';
export const TOGGLE_BLOGMODE = 'app/TOGGLE_BLOGMODE';
export const RECEIVE_FEATURE_FLAGS = 'app/RECEIVE_FEATURE_FLAGS';

export const defaultState = Map({
    loading: false,
    error: '',
    location: {},
    notifications: null,
    notificounters: Map({
        total: 0,
        feed: 0,
        reward: 0,
        send: 0,
        mention: 0,
        follow: 0,
        vote: 0,
        reply: 0,
        account_update: 0,
        message: 0,
        receive: 0,
    }),
    user_preferences: Map({
        locale: null,
        nsfwPref: 'warn',
        nightmode: false,
        blogmode: false,
        currency: 'USD',
    }),
    featureFlags: Map({}),
});

export default function reducer(state = defaultState, action = {}) {
    switch (action.type) {
        case '@@router/LOCATION_CHANGE':
            return state.set('location', { pathname: action.payload.pathname });
        case STEEM_API_ERROR:
            // Until we figure out how to better handle these errors, let em slide.
            // This action is the only part of the app that marks an error in state.app.error,
            // and the only part of the app which pays attn to this part of the state is in App.jsx.
            //return  state.set('error', action.error).set('loading', false);
            return state;
        case FETCH_DATA_BEGIN:
            return state.set('loading', true);
        case FETCH_DATA_END:
            return state.set('loading', false);
        case ADD_NOTIFICATION: {
            const n = {
                action: tt('g.dismiss'),
                dismissAfter: 10000,
                ...action.payload,
            };
            return state.update('notifications', s => {
                return s ? s.set(n.key, n) : OrderedMap({ [n.key]: n });
            });
        }
        case REMOVE_NOTIFICATION:
            return state.update('notifications', s =>
                s.delete(action.payload.key)
            );
        case UPDATE_NOTIFICOUNTERS: {
            if (action.payload) {
                const nc = action.payload;
                if (nc.follow > 0) {
                    nc.total -= nc.follow;
                    nc.follow = 0;
                }
                return state.set('notificounters', Map(nc));
            }
            return state;
        }
        case SET_USER_PREFERENCES:
            return state.set('user_preferences', Map(action.payload));
        case TOGGLE_NIGHTMODE:
            return state.setIn(
                ['user_preferences', 'nightmode'],
                !state.getIn(['user_preferences', 'nightmode'])
            );
        case TOGGLE_BLOGMODE:
            return state.setIn(
                ['user_preferences', 'blogmode'],
                !state.getIn(['user_preferences', 'blogmode'])
            );
        case RECEIVE_FEATURE_FLAGS:
            const newFlags = state.get('featureFlags')
                ? state.get('featureFlags').merge(action.flags)
                : Map(action.flags);
            return state.set('featureFlags', newFlags);
        default:
            return state;
    }
}

export const steemApiError = error => ({
    type: STEEM_API_ERROR,
    error,
});

export const fetchDataBegin = () => ({
    type: FETCH_DATA_BEGIN,
});

export const fetchDataEnd = () => ({
    type: FETCH_DATA_END,
});

export const addNotification = payload => ({
    type: ADD_NOTIFICATION,
    payload,
});

export const removeNotification = payload => ({
    type: REMOVE_NOTIFICATION,
    payload,
});

export const updateNotificounters = payload => ({
    type: UPDATE_NOTIFICOUNTERS,
    payload,
});

export const setUserPreferences = payload => ({
    type: SET_USER_PREFERENCES,
    payload,
});

export const toggleNightmode = () => ({
    type: TOGGLE_NIGHTMODE,
});

export const toggleBlogmode = () => ({
    type: TOGGLE_BLOGMODE,
});

export const receiveFeatureFlags = flags => ({
    type: RECEIVE_FEATURE_FLAGS,
    flags,
});

export const selectors = {
    getFeatureFlag: (state, flagName) =>
        state.getIn(['featureFlags', flagName], false),
};
