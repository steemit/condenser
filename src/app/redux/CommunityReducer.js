import { fromJS, Map, List } from 'immutable';

const GET_COMMUNITY_ROLES = 'community/GET_COMMUNITY_ROLES';
const GET_COMMUNITY_ROLES_PENDING = 'community/GET_COMMUNITY_ROLES_PENDING';
const SET_COMMUNITY_ROLES = 'community/SET_COMMUNITY_ROLES';
const GET_COMMUNITY_ROLES_ERROR = 'community/GET_COMMUNITY_ROLES_ERROR';

const GET_COMMUNITY_SUBSCRIBERS = 'community/GET_COMMUNITY_SUBSCRIBERS';
const GET_COMMUNITY_SUBSCRIBERS_PENDING =
    'community/GET_COMMUNITY_SUBSCRIBERS_PENDING';
const SET_COMMUNITY_SUBSCRIBERS = 'community/SET_COMMUNITY_SUBSCRIBERS';
const GET_COMMUNITY_SUBSCRIBERS_ERROR =
    'community/GET_COMMUNITY_SUBSCRIBERS_ERROR';

const UPDATE_USER_ROLE = 'community/UPDATE_USER_ROLE';
const SET_USER_ROLE_PENDING = 'community/SET_USER_ROLE_PENDING';
const APPLY_USER_ROLE = 'community/APPLY_USER_ROLE';

const defaultCommunityState = Map();

export default function reducer(state = defaultCommunityState, action) {
    const payload = action.payload;
    switch (action.type) {
        // Has Saga watcher.
        case GET_COMMUNITY_SUBSCRIBERS: {
            return state;
        }

        case GET_COMMUNITY_SUBSCRIBERS_PENDING: {
            const { community, pending } = payload;
            return state.setIn([community, 'listSubscribersPending'], pending);
        }

        case GET_COMMUNITY_SUBSCRIBERS_ERROR: {
            const { community, error } = payload;
            return state.setIn([community, 'listSubscribersError'], error);
        }

        case SET_COMMUNITY_SUBSCRIBERS: {
            const { community, subscribers } = payload;
            return state.setIn([community, 'subscribers'], subscribers);
        }

        // Has Saga watcher.
        case GET_COMMUNITY_ROLES: {
            return state;
        }

        case GET_COMMUNITY_ROLES_PENDING: {
            const { community, pending } = payload;
            return state.setIn([community, 'listRolesPending'], pending);
        }
        case SET_COMMUNITY_ROLES: {
            const { community, roles } = payload;
            state.setIn([community, 'roles'], fromJS(roles));
            return state.setIn([community, 'roles'], fromJS(roles));
        }

        case SET_USER_ROLE_PENDING: {
            const { community, pending } = payload;
            return state.setIn([community, 'updatePending'], pending);
        }

        case APPLY_USER_ROLE: {
            const { community } = payload;
            const index = state
                .getIn([community, 'roles'])
                .findIndex(r => r.get(0) === payload.account);

            if (index === -1) {
                return state.updateIn([community, 'roles'], List(), list => {
                    return list.withMutations(items => {
                        items.push(List([payload.account, payload.role]));
                    });
                });
            } else {
                return state.updateIn([community, 'roles'], List(), items =>
                    items.update(index, item =>
                        item.set(0, payload.account).set(1, payload.role)
                    )
                );
            }
        }

        default:
            return state;
    }
}

/**
    @arg {community: string} payload action payload.
*/
export const getCommunityRoles = payload => ({
    type: GET_COMMUNITY_ROLES,
    payload,
});

/**
    @arg boolean payload action payload.
*/
export const getCommunityRolesPending = payload => ({
    type: GET_COMMUNITY_ROLES_PENDING,
    payload,
});

export const getCommunityRolesError = payload => ({
    type: GET_COMMUNITY_ROLES_ERROR,
    payload,
});

export const setCommunityRoles = payload => ({
    type: SET_COMMUNITY_ROLES,
    payload,
});

/**
    @arg {community: string} payload action payload.
*/
export const getCommunitySubscribers = payload => ({
    type: GET_COMMUNITY_SUBSCRIBERS,
    payload,
});

/**
    @arg boolean payload action payload.
*/
export const getCommunitySubscribersPending = payload => ({
    type: GET_COMMUNITY_SUBSCRIBERS_PENDING,
    payload,
});

/**
    @arg error payload action payload.
*/
export const getCommunitySubscribersError = payload => ({
    type: GET_COMMUNITY_SUBSCRIBERS_ERROR,
    payload,
});

export const setCommunitySubscribers = payload => ({
    type: SET_COMMUNITY_SUBSCRIBERS,
    payload,
});

export const updateUserRole = payload => ({
    type: UPDATE_USER_ROLE,
    payload,
});
export const setUserRolePending = payload => ({
    type: SET_USER_ROLE_PENDING,
    payload,
});
export const applyUserRole = payload => ({
    type: APPLY_USER_ROLE,
    payload,
});
