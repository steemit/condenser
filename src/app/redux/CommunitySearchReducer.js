import { Map, List, fromJS } from 'immutable';

const COMMUNITY_SEARCH_DISPATCH = 'communitySearch/COMMUNITY_SEARCH_DISPATCH';
const COMMUNITY_SEARCH_PENDING = 'communitySearch/COMMUNITY_SEARCH_PENDING';
const COMMUNITY_SEARCH_ERROR = 'communitySearch/COMMUNITY_SEARCH_ERROR';
const COMMUNITY_SEARCH_RESULT = 'communitySearch/COMMUNITY_SEARCH_RESULT';

const defaultCommunitySearchState = Map({
    pending: false,
    error: false,
    result: List([]),
});

export default function reducer(state = defaultCommunitySearchState, action) {
    const payload = action.payload;

    switch (action.type) {
        // Has a saga watcher.
        case COMMUNITY_SEARCH_DISPATCH: {
            return state;
        }
        case COMMUNITY_SEARCH_PENDING: {
            const { pending } = payload;
            return state.setIn(['pending'], pending);
        }
        case COMMUNITY_SEARCH_ERROR: {
            const { error } = payload;
            return state.setIn(['error'], error);
        }
        case COMMUNITY_SEARCH_RESULT: {
            // TODO: format / parse the search result and put it into state
            return state;
        }
        default:
            return state;
    }
}

export const communitySearch = payload => ({
    type: COMMUNITY_SEARCH_DISPATCH,
    payload,
});
export const communitySearchPending = payload => ({
    type: COMMUNITY_SEARCH_PENDING,
    payload,
});
export const communitySearchError = payload => ({
    type: COMMUNITY_SEARCH_PENDING,
    payload,
});

export const communitySearchResult = payload => ({
    type: COMMUNITY_SEARCH_RESULT,
    payload,
});
