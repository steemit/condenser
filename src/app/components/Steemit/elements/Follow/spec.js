import React from 'react';
import { mount, configure, shallow } from 'enzyme';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import Follow from './index';
import rootReducer from 'app/redux/RootReducer';
import Adapter from 'enzyme-adapter-react-15';

const store = createStore(rootReducer);

configure({ adapter: new Adapter() });

describe('<Follow />', () => {
    const wrapper = shallow(
        <Provider store={store}>
            <Follow />
        </Provider>
    );
    const container = wrapper.instance();

    it('renders without crashing', () => {
        expect(wrapper).toBeTruthy();
        expect(container).toBeTruthy();
        expect(wrapper).toMatchSnapshot();
    });
});
