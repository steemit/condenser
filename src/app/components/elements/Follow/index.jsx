import React, { PropTypes } from 'react';
import { connect } from 'react-redux';
import LoadingIndicator from 'app/components/elements/LoadingIndicator';
import shouldComponentUpdate from 'app/utils/shouldComponentUpdate';
import * as transactionActions from 'app/redux/TransactionReducer';
import * as userActions from 'app/redux/UserReducer';
import { Set, Map } from 'immutable';
import tt from 'counterpart';

const { string, bool, any } = PropTypes;

export default class Follow extends React.Component {
    static propTypes = {
        following: string,
        follower: string, // OPTIONAL default to current user
        showFollow: bool,
        showMute: bool,
        fat: bool,
        children: any,
        showLogin: React.PropTypes.func.isRequired,
    };

    static defaultProps = {
        showFollow: true,
        showMute: true,
        fat: false,
    };

    constructor(props) {
        super();
        this.state = {};
        this.initEvents(props);
        this.followLoggedOut = this.followLoggedOut.bind(this);
        this.shouldComponentUpdate = shouldComponentUpdate(this, 'Follow');
    }

    componentWillUpdate(nextProps) {
        this.initEvents(nextProps);
    }

    initEvents(props) {
        const { updateFollow, follower, following } = props;
        const upd = type => {
            if (this.state.busy) return;
            this.setState({ busy: true });
            const done = () => {
                this.setState({ busy: false });
            };
            updateFollow(follower, following, type, done);
        };
        this.follow = () => {
            upd('blog');
        };
        this.unfollow = () => {
            upd();
        };
        this.ignore = () => {
            upd('ignore');
        };
        this.unignore = () => {
            upd();
        };
    }

    followLoggedOut(e) {
        // close author preview if present
        const author_preview = document.querySelector('.dropdown-pane.is-open');
        if (author_preview) author_preview.remove();
        // resume authenticate modal
        this.props.showLogin(e);
    }

    render() {
        const { loading } = this.props;
        if (loading)
            return (
                <span>
                    <LoadingIndicator /> {tt('g.loading')}&hellip;
                </span>
            );
        if (loading !== false) {
            // must know what the user is already following before any update can happen
            return <span />;
        }

        const { follower, following } = this.props; // html
        // Show follow preview for new users
        if (!follower || !following)
            return (
                <span>
                    <label
                        className="button slim hollow secondary"
                        onClick={this.followLoggedOut}
                    >
                        {tt('g.follow')}
                    </label>
                </span>
            );
        // Can't follow or ignore self
        if (follower === following) return <span />;

        const { followingWhat } = this.props; // redux
        const { showFollow, showMute, fat, children } = this.props; // html
        const { busy } = this.state;

        const cnBusy = busy ? 'disabled' : '';
        const cnActive = 'button' + (fat ? '' : ' slim');
        const cnInactive = cnActive + ' hollow secondary ' + cnBusy;
        return (
            <span>
                {showFollow &&
                    followingWhat !== 'blog' && (
                        <label className={cnInactive} onClick={this.follow}>
                            {tt('g.follow')}
                        </label>
                    )}

                {showFollow &&
                    followingWhat === 'blog' && (
                        <label className={cnInactive} onClick={this.unfollow}>
                            {tt('g.unfollow')}
                        </label>
                    )}

                {showMute &&
                    followingWhat !== 'ignore' && (
                        <label className={cnInactive} onClick={this.ignore}>
                            {tt('g.mute')}
                        </label>
                    )}

                {showMute &&
                    followingWhat === 'ignore' && (
                        <label className={cnInactive} onClick={this.unignore}>
                            {tt('g.unmute')}
                        </label>
                    )}

                {children && <span>&nbsp;&nbsp;{children}</span>}
            </span>
        );
    }
}

const emptyMap = Map();
const emptySet = Set();

module.exports = connect(
    (state, ownProps) => {
        let { follower } = ownProps;
        if (!follower) {
            const current_user = state.user.get('current');
            follower = current_user ? current_user.get('username') : null;
        }

        const { following } = ownProps;
        const f = state.global.getIn(
            ['follow', 'getFollowingAsync', follower],
            emptyMap
        );

        // the line below was commented out by val - I think it's broken so sometimes the loading indicator is shown forever
        // const loading = f.get('blog_loading', false) || f.get('ignore_loading', false)
        const loading = false;

        const followingWhat = f.get('blog_result', emptySet).contains(following)
            ? 'blog'
            : f.get('ignore_result', emptySet).contains(following)
              ? 'ignore'
              : null;

        return {
            follower,
            following,
            followingWhat,
            loading,
        };
    },
    dispatch => ({
        updateFollow: (follower, following, action, done) => {
            const what = action ? [action] : [];
            const json = ['follow', { follower, following, what }];
            dispatch(
                transactionActions.broadcastOperation({
                    type: 'custom_json',
                    operation: {
                        id: 'follow',
                        required_posting_auths: [follower],
                        json: JSON.stringify(json),
                    },
                    successCallback: done,
                    errorCallback: done,
                })
            );
        },
        showLogin: e => {
            if (e) e.preventDefault();
            dispatch(userActions.showLogin());
        },
    })
)(Follow);
