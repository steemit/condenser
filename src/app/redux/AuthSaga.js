import {takeEvery} from 'redux-saga';
import {call, put, select} from 'redux-saga/effects';
import {Set, Map, fromJS, List} from 'immutable'
import {getAccount} from 'app/redux/SagaShared'
import {PrivateKey} from 'steem/lib/auth/ecc';
import {api} from 'steem';

// operations that require only posting authority
const postingOps = Set(`vote, comment, delete_comment, custom_json, claim_reward_balance`.trim().split(/,\s*/))

export const authWatches = [
    watchForAuth
]

function* watchForAuth() {
    yield* takeEvery('user/ACCOUNT_AUTH_LOOKUP', accountAuthLookup);
}

export function* accountAuthLookup({payload: {account, private_keys, login_owner_pubkey}}) {
    account = fromJS(account)
    private_keys = fromJS(private_keys)
    // console.log('accountAuthLookup', account.name)
    const stateUser = yield select(state => state.user)
    let keys
    if (private_keys)
        keys = private_keys
    else
        keys = stateUser.getIn(['current', 'private_keys'])

    if (!keys || !keys.has('posting_private')) return
    const toPub = k => k ? k.toPublicKey().toString() : '-'
    const posting = keys.get('posting_private')
    const active = keys.get('active_private')
    const memo = keys.get('memo_private')
    const auth = {
        posting: posting ? yield authorityLookup(
            {pubkeys: Set([toPub(posting)]), authority: account.get('posting'), authType: 'posting'}) : 'none',
        active: active ? yield authorityLookup(
            {pubkeys: Set([toPub(active)]), authority: account.get('active'), authType: 'active'}) : 'none',
        owner: 'none',
        memo: account.get('memo_key') === toPub(memo) ? 'full' : 'none'
    }
    const accountName = account.get('name')
    const pub_keys_used = {posting: toPub(posting), active: toPub(active), owner: login_owner_pubkey};
    yield put({type: 'user/SET_AUTHORITY', payload: {accountName, auth, pub_keys_used}});
}

/**
    @arg {object} data
    @arg {object} data.authority Immutable Map blockchain authority
    @arg {object} data.pubkeys Immutable Set public key strings
    @return {string} full, partial, none
*/
function* authorityLookup({pubkeys, authority, authType}) {
    return yield call(authStr, {pubkeys, authority, authType})
}

function* authStr({pubkeys, authority, authType, recurse = 1}) {
    const t = yield call(threshold, {pubkeys, authority, authType, recurse})
    const r = authority.get('weight_threshold')
    return t >= r ? 'full' : t > 0 ? 'partial' : 'none'
}

export function* threshold({pubkeys, authority, authType, recurse = 1}) {
    if (!pubkeys.size) return 0
    let t = pubkeyThreshold({pubkeys, authority})
    const account_auths = authority.get('account_auths')
    const aaNames = account_auths.map(v => v.get(0), List())
    if (aaNames.size) {
        const aaAccounts = yield api.getAccountsAsync(aaNames)
        const aaThreshes = account_auths.map(v => v.get(1), List())
        for (let i = 0; i < aaAccounts.size; i++) {
            const aaAccount = aaAccounts.get(i)
            t += pubkeyThreshold({authority: aaAccount.get(authType), pubkeys})
            if (recurse <= 2) {
                const auth = yield call(authStr,
                    {authority: aaAccount, pubkeys, recurse: ++recurse})
                if (auth === 'full') {
                    const aaThresh = aaThreshes.get(i)
                    t += aaThresh
                }
            }
        }
    }
    return t
}

function pubkeyThreshold({pubkeys, authority}) {
    let available = 0
    const key_auths = authority.get('key_auths')
    key_auths.forEach(k => {
        if (pubkeys.has(k.get(0))) {
            available += k.get(1)
        }
    })
    return available
}

export function* findSigningKey({opType, username, password}) {
    let authTypes
    if (postingOps.has(opType))
        authTypes = 'posting, active'
    else
        authTypes = 'active, owner'
    authTypes = authTypes.split(', ')

    const currentUser = yield select(state => state.getIn(['user', 'current']))
    const currentUsername = currentUser && currentUser.get('username')

    username = username || currentUsername
    if (!username) return null

    const private_keys = currentUsername === username ? currentUser.get('private_keys') : Map()

    const account = yield call(getAccount, username);
    if (!account) throw new Error('Account not found')

    for (const authType of authTypes) {
        let private_key
        if (password) {
            try {
                private_key = PrivateKey.fromWif(password)
            } catch (e) {
                private_key = PrivateKey.fromSeed(username + authType + password)
            }
        } else {
            if(private_keys)
                private_key = private_keys.get(authType + '_private')
        }
        if (private_key) {
            const pubkey = private_key.toPublicKey().toString()
            const pubkeys = Set([pubkey])
            const authority = account.get(authType)
            const auth = yield call(authorityLookup, {pubkeys, authority, authType})
            if (auth === 'full') return private_key
        }
    }
    return null
}

// function isPostingOnlyKey(pubkey, account) {
//     // TODO Support account auths
//     // yield put(g.actions.authLookup({account, pubkeys: pubkey})
//     // authorityLookup({pubkeys, authority: Map(account.posting), authType: 'posting'})
//     for (const p of account.posting.key_auths) {
//         if (pubkey === p[0]) {
//             if (account.active.account_auths.length || account.owner.account_auths.length) {
//                 console.log('UserSaga, skipping save password, account_auths are not yet supported.')
//                 return false
//             }
//             for (const a of account.active.key_auths)
//                 if (pubkey === a[0]) return false
//             for (const a of account.owner.key_auths)
//                 if (pubkey === a[0]) return false
//             return true
//         }
//     }
//     return false
// }
