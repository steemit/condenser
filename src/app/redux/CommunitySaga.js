import { call, put, takeEvery, select } from 'redux-saga/effects';
import * as reducer from 'app/redux/CommunityReducer';
import { getCommunity } from 'app/redux/FetchDataSaga';
import { callBridge } from 'app/utils/steemApi';
import * as transactionActions from './TransactionReducer';

export const communityWatches = [
    takeEvery('community/GET_COMMUNITY_ROLES', getCommunityRoles),
    takeEvery('community/GET_COMMUNITY_SUBSCRIBERS', getCommunitySubscribers),
    takeEvery('community/UPDATE_USER_ROLE', updateUserRole),
];

export function* getCommunityRoles(action) {
    const community = action.payload;
    yield put(reducer.getCommunityRolesPending({ community, pending: true }));
    try {
        const roles = yield call(callBridge, 'list_community_roles', {
            community,
        });
        yield call(getCommunity, action);
        yield put(reducer.setCommunityRoles({ community, roles }));
    } catch (error) {
        yield put(reducer.getCommunityRolesError({ community, error }));
    }
    yield put(reducer.getCommunityRolesPending({ community, pending: false }));
}

export function* getCommunitySubscribers(action) {
    const community = action.payload;
    yield put(
        reducer.getCommunitySubscribersPending({ community, pending: true })
    );
    try {
        const subscribers = yield call(callBridge, 'list_subscribers', {
            community,
        });
        yield call(getCommunity, action);
        yield put(reducer.setCommunitySubscribers({ community, subscribers }));
    } catch (error) {
        yield put(reducer.getCommunitySubscribersError({ community, error }));
    }
    yield put(
        reducer.getCommunitySubscribersPending({ community, pending: false })
    );
}

export function* updateUserRole(action) {
    const { community } = action.payload;
    yield put(reducer.setUserRolePending({ community, pending: true }));
    try {
        const username = yield select(state =>
            state.user.getIn(['current', 'username'])
        );

        yield put(
            transactionActions.broadcastOperation({
                type: 'custom_json',
                operation: {
                    id: 'community',
                    required_posting_auths: [username],
                    json: JSON.stringify(['setRole', action.payload]),
                },
                //successCallback,
                //errorCallback,
            })
        );

        yield put(reducer.applyUserRole(action.payload));
    } catch (error) {
        console.log('update user error', error);
    }
    yield put(reducer.setUserRolePending({ community, pending: false }));
}
