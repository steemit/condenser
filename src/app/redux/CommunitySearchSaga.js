import { put, takeEvery, call } from 'redux-saga/effects';
import * as reducer from 'app/redux/CommunitySearchReducer';

export const searchWatches = [
    takeEvery('communitySearch/COMMUNITY_SEARCH_DISPATCH', communitySearchSaga),
];

export function* communitySearchSaga(action) {
    // const { q, s, scroll_id } = action.payload;
    yield put(reducer.communitySearchPending({ pending: true }));
    try {
        // TODO: Search Call goes here, waiting on API endpoint
        //yield put(reducer.searchResult({ , append }));
    } catch (error) {
        console.log('Search error', error);
        yield put(reducer.communitySearchError({ error }));
    }
    yield put(reducer.communitySearchPending({ pending: false }));
}
