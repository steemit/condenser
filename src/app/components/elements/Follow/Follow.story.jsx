import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, boolean } from '@storybook/addon-knobs';
import rootReducer from 'app/redux/RootReducer';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import Follow from './index';
import { Center } from '../Tooltip.story';

const store = createStore(rootReducer);

storiesOf('Elements', module)
    .addDecorator(withKnobs)
    .addDecorator(getStory => <Provider store={store}>{getStory()}</Provider>)
    .add('Follow', () => (
        <Center>
            <Follow
                className="float-right"
                follower={'maitland'}
                following={'Carol'}
                what="blog"
                showFollow={boolean('Show Follow?', true)}
                showMute={boolean('Show Mute?', true)}
            />
        </Center>
    ));
