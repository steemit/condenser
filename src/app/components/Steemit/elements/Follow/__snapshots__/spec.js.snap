// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Follow /> renders without crashing 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Provider
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
  >
    <Connect(Follow) />
  </Provider>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {},
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {},
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
}
`;
